pym/bob/layers.py

```diff
@@ -0,0 +1,193 @@
+import datetime
+import os
+from .errors import BuildError
+from .invoker import CmdFailedError, InvocationError, Invoker
+from .state import BobState
+from .input import RecipeSet, Scm, YamlCache
+from .utils import INVALID_CHAR_TRANS
+from .tty import DEBUG, EXECUTED, INFO, NORMAL, IMPORTANT, SKIPPED, WARNING, log
+
+class LayerStepSpec:
+    def __init__(self, path):
+        self.__path = path
+
+    @property
+    def workspaceWorkspacePath(self):
+        return self.__path
+
+    @property
+    def env(self):
+        return {}
+
+    @property
+    def envWhiteList(self):
+        return []
+
+class Layer:
+    def __init__(self, name, root, recipes, yamlCache, scm=None):
+        self.__name = name
+        self.__root = root
+        self.__recipes = recipes
+        self.__yamlCache = yamlCache
+        self.__subLayers = []
+        self.__scm = scm
+        self.__layerDir = os.path.join(self.__root, "layers", self.__name) if len(self.__name) else self.__root
+
+    async def __checkoutTask(self, verbose):
+        if self.__scm is None:
```

Comment:
So we don't move obsolete layers into the attic? This is different compared to normal checkouts in recipes...

Reply:
I think this test is obsolete and can be removed. In `Layers::parse` a layer without a scm is already skipped.

Anyway - I think this is a edge case and I'm not sure how tho handle cases where a manager-layer is tured back into a external-layer. This not easily possible and would require multiple steps anyway. A simple `git pull --recures-submodules` would fail as the submodule-directoy already exists? So the user would need to update the recipes, run `bob layers update` to move the layer into attic and than update the submodules to make this work? I think in this case some communication to the team is needed in any case and than you can also advise them to move the existing layer manually?

Reply:
The usual checkout logic has the following rules:

* A new checkout checks that there is not already a file or directory at the checkout location.
* Only checkouts are moved to the attic when their update fails.
* Obsolete checkouts are always moved to the attic.

If possible, I would like to have the same behaviour for managed layers.

If I understand your example correctly you have a project with submodules and now, after a git-pull the submodules is gone and the layer(s) is/are managed by Bob itself? I'm afraid this requires some communication. Git is not handling vanishing sub-layers well (and neither does Bob). Git will just abandon the submodule and the managed layers logic will bail out because the new checkout will collide with the abandoned submodule.

I think that letting the user sort that out is reasonable. Bob tries to checkout a layer and finds a directory there that is unknown. The only sensible default is to bail out. I guess this won't happen often.

Reply:
```quote
Obsolete checkouts are always moved to the attic.
```

Question is how `obsolete` is defined here. If I do no longer depend on a package the checkout for this is not moved to attic. The checkout for this is no longer run so the sources were kept in the workspace.

So if a layer is removed from the `config.yaml` would it become obsolete? As of today it's fine to have more layers in the `layers` directory than referenced by the `config.yaml`.

Or if the 'managed'-foo layer
```yaml
layers:
 - name: foo
   scm: git
   url: ....
```
becomes unmanaged:
```yaml
layers:
 - foo
```
is it now obsolete? It's still used... :shrug:

Reply:
I would argue that, if a layer was checked out by Bob and the new version of `config.yaml` does not have a checkout, it should be moved to the attic. I see the project with the SCM layers like the source workspace of a recipe. When its updated, it should reflect the state of the recipe. In case of layers, its the state of `config.yaml` but the same logic should apply. Its the principle of consistency and least surprise.

Keeping the layer around might be confusing because the recipes are there but simply unused. Sure, it's the current behaviour for unmanaged layers but if Bob brings the layer into existence, it should probably also be moved away when being removed from the configuration. Additionally, what should happen if the next `config.yaml`, after the previous layer was abandoned, checks out a completely different layer `foo`? Maybe even by a different SCM? That must fail because there is already something at the checkout location.

So in your example, it would be moved to the attic. But that's interesting because I guess the `git pull` of the project will probably have failed. Then the user will take care of `foo` himself (delete/move away) and do the git pull again. Now, given the above logic, Bob will move the `foo` directory into the attic. This is clearly not what we want. So we need to handle the managed->unmanaged transition specially.

In summary, these may be the coarse rules:

1. New managed layers are not checked out into existing directories. Like normal SCMs, the directory must not exist on the initial checkout.
2. The transition of a layer from unmanaged to managed uses the same rules. That is, the user is responsible to remove the directory at the place of the newly managed layer first.
3. As always, we try to update in-place and if this fails move to the attic. I guess we want to give `bob layers` some switches to be able to tune the behaviour.
4. If a managed layer is removed, the layer is moved into the attic.
5. A layer that is transitioning from managed to unmanaged is abandoned. We don't want to move something into the attic that has been placed there by the user. This is the only deviation from the usual source workspace update rules.

Reply:
Thanks! I updated the code accordingly and added some tests for the different situations.